hawc/apps/epi/api.py

```diff
@@ -87,37 +100,425 @@ def result_heatmap(self, request, pk):
+                    ).delete()
+
+                criteria_ids = self.request.data[data_key]
+
+                for criteria_id in criteria_ids:
+                    dynamic_obj = {
+                        "criteria_type": type_code,
+                        "criteria": criteria_id,
+                        "study_population": study_population_id,
+                    }
+                    inserts.append(dynamic_obj)
+
+        # ...and save any new ones
+        if len(inserts) > 0:
+            serializer = serializers.SimpleStudyPopulationCriteriaSerializer(
+                data=inserts, many=True
+            )
+            serializer.is_valid(raise_exception=True)
+            serializer.save()
+
+    def get_queryset(self, *args, **kwargs):
+        return self.model.objects.all()
+
+    def handle_criteria(self, request, during_create):
+        # first - what assessment are we working in?
+        assessment_id = None
+        if during_create:
+            if "study" in request.data:
+                study_id = request.data["study"]
+
+                try:
+                    study = Study.objects.get(id=study_id)
+                    assessment_id = study.get_assessment().id
+                except ObjectDoesNotExist:
+                    raise ValidationError({"study": "Invalid study id"})
+            else:
+                raise ValidationError({"study": "No study id supplied"})
+        else:
+            assessment_id = self.get_object().get_assessment().id
+
+        if assessment_id is None:
+            return
+
+        # and now do some conversions/creations
+        for cc in self.criteria_categories:
+            data_key = cc[0]
+
+            # client can supply an id, or the name of the criteria entry (and then we'll look it up for them - or create it if needed)
+            if data_key in request.data:
+                data_probe = request.data[data_key]
+
+                fixed = []
+                for el in data_probe:
+                    if type(el) is str:
+                        try:
+                            criteria = models.Criteria.objects.get(
+                                description=el, assessment_id=assessment_id
+                            )
+                            fixed.append(criteria.id)
+                        except ObjectDoesNotExist:
+                            # allow creation of criteria as part of the request
+                            criteria_serializer = serializers.CriteriaSerializer(
+                                data={"description": el, "assessment": assessment_id}
+                            )
+                            criteria_serializer.is_valid(raise_exception=True)
+                            criteria = criteria_serializer.save()
+                            fixed.append(criteria.id)
+                    else:
+                        fixed.append(el)
+
+                request.data[data_key] = fixed
+
+    def perform_create(self, serializer):
```

Comment:
wrap this and all others in a transaction in case we get in a bad state and need to revert https://docs.djangoproject.com/en/3.1/topics/db/transactions/#django.db.transaction.atomic

Please do throughout the api.py